File: src/main/problems/java/Nr102.java
package main.problems.java;

import java.util.ArrayList;
import java.util.List;

import main.util.java.TreeNode;

public class Nr102 {
    public List<List<Integer>> levelOrder(TreeNode root) {
        return createLevels(root, new ArrayList<>(), 0);
    }

    private List<List<Integer>> createLevels(TreeNode root, List<List<Integer>> levels, int depth) {
        if (root == null) {
            return levels;
        }
        if (levels.size() - 1 < depth) {
            levels.add(new ArrayList<>());
        }
        levels.get(depth).add(root.val);
        createLevels(root.left, levels, depth + 1);
        createLevels(root.right, levels, depth + 1);
        return levels;
    }
}

File: src/main/problems/ruby/Nr102.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[][]}
def level_order(root, levels = [], level = 0)
    return [] unless root

    levels << [] unless levels[level]
    levels[level] << root.val
    level_order(root.left, levels, level + 1)
    level_order(root.right, levels, level + 1)
    levels
end